### PR TITLE
Update data check actions API to return options instead of actions and add functionality to suggest and take action on columns with null values

### LINKED ISSUE #3013: Support per-column imputation via data checks / data check actions
Right now, our data checks will suggest a IMPUTE_COLS action which is only used to impute the target column. We should also support an action that suggests per-column imputation for features with missing values. This will likely require a change in the actions API to store imputation strategy, fill value, and possibly other parameters associated with each column.

### LINKED ISSUE #3116: Update data checks' `validate` method to return actions under new API
We want to support the idea of returning action options from `validate()`, rather than the `actions` themselves.

Right now, we return:
```
data_check = InvalidTargetDataCheck("binary", "Log Loss Binary")
data_checks_output = data_check.validate(None, y)
>>> [{"code": "IMPUTE_COL", 
      "metadata": {
          "impute_strategy": "most_frequent",
          "is_target": True,
          "rows": None,
          "columns": None
      }
  }
]
```

Completion of this issue would update `validate` to look something like this instead: 
```
data_check = InvalidTargetDataCheck("binary", "Log Loss Binary")
data_checks_output = data_check.validate(None, y)
>>>   
{
  "warnings": [...],
  "errors": [...],
  "actions": {
    "actions_list": [
{
  "data_check_name"
  "action_code": IMPUTE_COLS,
  "parameters": {
    "impute_strategies": {
      "parameter_type": "column",
      "columns": {
        "a": {
          "impute_strategy”: {"categories": ["mean", "mode"], "type": "category", "default_value": "mean"},
          "constant_fill_value": {"type": "float", "default_value": 0}
        }
        "b": {
          "impute_strategy”: {"categories": ["mean", "mode"], "type": "category", "default_value": "mean"},
          "constant_fill_value": {"type": "float", "default_value": 0}
        }
        "c": {
          "impute_strategy”: {"categories": ["mean", "mode"], "type": "category", "default_value": "mean"},
          "constant_fill_value": {"type": "float", "default_value": 0}
        }
        ...
      }
    }
    "constant_fill_value": {
        "type": "float"
    }
  }
  "is_target": False,
  "rows": None,
  "columns": ["a", "b", "c", "d", "e"]
}
    ],
    "default_action": 0 # index to action that is considered the default action to take
}
```

### LINKED ISSUE #3144: Update `HighlyNullDataCheck` to detect when columns have any null values
Add new `NullDataCheck` data check that will warn users of any null values in their columns and return an action to impute columns with null values.

### LINKED ISSUE #3240: NaturalLanguageDataCheck should raise an error for highly null columns and a warning for columns with nulls
As a user of `NaturalLanguageDataCheck`, I would like for this data check to
- Raise a `DataCheckError` if a natural language column has more NaNs than a given threshold
- Rase a `DataCheckWarning` if a natural language column has NaNs but is below the given threshold

This is basically changing the `NaturalLanguageDataCheck` to behave like the `NullDataCheck` (formerly `HighlyNullDataCheck`)
#### Code Example

```python
X = pd.DataFrame({"some_null": ["Natural Language", None, "Is great"],
                                  "highly_null": ["I am happy", None, None]})
data_check = NaturalLanguageNaNDataCheck(pct_null_col_threshold=0.5)
validation = data_check.validate(X)
```

Closing Criteria:
- DataCheckWarning for `some_null`
- DataCheckError for `highly_null`
- Recommended action of dropping `highly_null`

### LINKED ISSUE #3243: Update data check's ``validate()`` method to return a flat hierarchy of messages
Per discussion with team, perhaps we should remove the dictionary output and instead, update the data check's ``validate()`` method to return a flat hierarchy of messages.

(Essentially, reverting https://github.com/alteryx/evalml/issues/1325)

### PR DESCRIPTION
**Pull request by angela97lin**
Closes #3116, closes #3013, closes #3144, closes #3243, closes #3240

Everything in this branch should have been previously reviewed; I just didn't want to continuously break and fix any code that was using the intermediate API changes as I broke down issues into smaller (and hopefully easier to review) PRs.

### CODE REVIEW

**Comment by angela97lin** [comment]
EDIT: oops, too excited. Meant to ask for review for https://github.com/alteryx/evalml/pull/3237 first. Once that's merged in, I'll open this officially :)

File: evalml/tests/data_checks_tests/test_data_check_action_option.py
Comment (chukarsten) [inline evalml/tests/data_checks_tests/test_data_check_action_option.py:430]: Yikes!

**Review by chukarsten — APPROVED** [review]
This looks good, Angela!  I'm sorry I wasn't able to post some of the reviews I did of your constituent PRs, I had a few pending submission when they got merged.  But they were all approvals anyways.  This looks good to me and is great work.  Way to really stick it through...this one must have been exhausting.

File: evalml/data_checks/null_data_check.py
Comment (freddyaboulton) [inline evalml/data_checks/null_data_check.py:220]: I think we can improve this implementation. Right now we do two scans of the data to determine the highly null columns and the columns with any nulls. Doing so also involves creating a new copy of the data without NaturalLanguage/DateTime. 

I think if we refactor `get_null_column_information` to return {col_name: % nulls} we only have to do one scan without creating another copy to determine which are highly null and which are null. 

Let's file an issue to do that in a separate PR.

**Review by freddyaboulton — APPROVED** [review]
@angela97lin This looks good to me! Thanks for all the work in this refactor.

I skimmed this PR as I reviewed all the sub PRs. Only one thing stood out to me.

File: evalml/data_checks/null_data_check.py
Comment (angela97lin) [reply evalml/data_checks/null_data_check.py:220]: You're right, thank you for pointing out that nuance! I filed #3273 to track this :)

**Review by ParthivNaresh — APPROVED** [review]
Fantastic work on this, I can't wait for it to be merged. And great coordination on splitting it up into manageable PRs first!

File: evalml/pipelines/utils.py
Comment (eccabay) [inline evalml/pipelines/utils.py:755]: With the merge of #3267, we don't need the `impute_all` parameter any more!

**Review by eccabay — APPROVED** [review]
Fantastic! Just left a few nits

File: docs/source/release_notes.rst
Comment (angela97lin) [reply docs/source/release_notes.rst:43]: @eccabay Haha yes! I put it in the wrong place before and am sneaking this in here :)

File: evalml/pipelines/utils.py
Comment (angela97lin) [reply evalml/pipelines/utils.py]: That's true! Thank you @eccabay 🙏

**Pull request merged**
